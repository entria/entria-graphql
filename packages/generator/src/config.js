/* @flow */
import path from 'path';
import merge from 'lodash.merge';
import fs from 'fs';
import pkgDir from 'pkg-dir';

const rootPath = pkgDir.sync('.') || '.';

let cacheConfig = null;

/**
 * Parse `.graphqlrc` config file and retrieve its contents
 * @param filePath {string} The path of the config file
 * @returns {*}
 */
const parseConfigFile = (filePath: string) => {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  const directories = Object.keys(config.directories).reduce((data, directory) => {
    if (directory === DIRECTORY_TYPE.SOURCE) {
      return {
        ...data,
        [directory]: `${rootPath}/${config.directories[directory]}`,
        [DIRECTORY_TYPE.SRC]: config.directories.source,
      };
    }

    return {
      ...data,
      [directory]: `${config.directories.source}/${config.directories.module}/${config.directories[directory]}`,
    };
  }, {});

  return {
    ...config,
    directories: {
      ...config.directories,
      ...directories,
    },
  };
};

export const DIRECTORY_TYPE = {
  SOURCE: 'source',
  SRC: 'src',
  MODULE: 'module',
  CONNECTION: 'connection',
  LOADER: 'loader',
  MODEL: 'model',
  MUTATION: 'mutation',
  MUTATION_TEST: 'mutation_test',
  TYPE: 'type',
  TYPE_TEST: 'type_test',
};

type DirectoryType = $Values<typeof DIRECTORY_TYPE>;
type Directories = {
  source: string,
  module: string,
  connection: string,
  loader: string,
  model: string,
  mutation: string,
  mutation_test: string,
  type: string,
  type_test: string,
  interface: string,
};

type Files = {
  schema: string,
};

type Config = {
  directories: Directories,
  files: Files,
}
/**
 * Get the `.graphqlrc` config file
 * @returns {object} The content of the config
 */
export const getCreateGraphQLConfig = (): Config => {
  // if (cacheConfig) return cacheConfig;

  // Use default config
  const defaultFilePath = path.resolve(`${__dirname}/graphqlrc.json`);

  const config = parseConfigFile(defaultFilePath);

  try {
    // Check if there is a `.graphqlrc` file in the root path
    const customConfig = parseConfigFile(`${rootPath}/.graphqlrc`);

    merge(config, customConfig);

    cacheConfig = config;

    // If it does, extend default config with it, so if the custom config has a missing line
    // it won't throw errors
    return config;
  } catch (err) {
    cacheConfig = config;
    // Return the default config if the custom doesn't exist
    return config;
  }
};

/**
 * Get a directory from the configuration file
 * @param directory {string} The name of the directory, e.g. 'source'/'mutation'
 * @returns {string} The directory path
 */
export const getConfigDir = (directory: DirectoryType) => getCreateGraphQLConfig().directories[directory];
