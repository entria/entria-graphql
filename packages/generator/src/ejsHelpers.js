/* @flow */
/**
 * Uppercase the first letter of a text
 * @param text {string}
 * @returns {string}
 */
export const uppercaseFirstLetter = (text: string) => `${text.charAt(0).toUpperCase()}${text.slice(1)}`;
/**
 * Camel cases text
 * @param text {string} Text to be camel-cased
 * @returns {string} Camel-cased text
 */
export const camelCaseText = (text: string) =>
  text.replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, (match, index) => {
    if (+match === 0) {
      return '';
    }

    return index === 0 ? match.toLowerCase() : match.toUpperCase();
  });
