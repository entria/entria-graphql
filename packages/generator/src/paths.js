/* @flow */
import path from 'path';

export const getModulePath = (source: string, moduleName: string) => {
  return path.join(source, moduleName);
};

export const getTestPath = (source: string) => {
  return path.join(source, '__tests__');
};

export const getMutationPath = (source: string) => {
  return path.join(source, 'mutation');
};
