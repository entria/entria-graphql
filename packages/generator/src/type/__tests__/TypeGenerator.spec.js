import helper from 'yeoman-test';
import assert from 'yeoman-assert';
import path from 'path';
import fs from 'fs-extra';

import {
  getFileContent,
  getFixturePath,
} from '../../../test/helpers';

import { getConfigDir } from '../../config';
import { getModulePath, getTestPath } from '../../paths';
import { uppercaseFirstLetter } from '../../ejsHelpers';

const typeGenerator = path.join(__dirname, '..');

it('generate a type', async () => {
  const moduleName = 'example';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(typeGenerator)
    .withArguments(name)
    .toPromise();

  const destinationDir = getConfigDir('type');
  const destinationTestDir = getConfigDir('type_test');

  const modulePath = getModulePath(destinationDir, moduleName);
  const testPath = getTestPath(modulePath);

  const typeFilepath = path.join(modulePath, `${name}Type.js`);
  const typeTestFilepath = path.join(testPath, `${name}Type.spec.js`);

  assert.file([
    typeFilepath,
    typeTestFilepath,
  ]);

  const files = {
    type: getFileContent(`${folder}/${destinationDir}/example/ExampleType.js`),
    typeTest: getFileContent(`${folder}/${destinationTestDir}/example/__tests__/ExampleType.spec.js`),
  };

  expect(files).toMatchSnapshot();
});

it('generate a type with schema', async () => {
  const moduleName = 'post';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(typeGenerator)
    .inTmpDir(dir =>
      fs.copySync(
        getFixturePath(name),
        path.join(dir, 'src/modules/post/PostModel.js'),
      ),
    )
    .withArguments('Post Post')
    .toPromise();

  const destinationDir = getConfigDir('type');

  const modulePath = getModulePath(destinationDir, moduleName);
  const testPath = getTestPath(modulePath);

  const typeFilepath = path.join(modulePath, `${name}Type.js`);
  const typeTestFilepath = path.join(testPath, `${name}Type.spec.js`);

  assert.file([
    typeFilepath,
    typeTestFilepath,
  ]);

  const files = {
    type: getFileContent(path.join(folder, typeFilepath)),
    typeTest: getFileContent(path.join(folder, typeTestFilepath)),
  };

  expect(files).toMatchSnapshot();
});

it('generate a type with schema and without timestamps', async () => {
  const moduleName = 'user';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(typeGenerator)
    .inTmpDir(dir =>
      fs.copySync(
        getFixturePath('User'),
        path.join(dir, 'src/modules/user/UserModel.js'),
      ),
    )
    .withArguments('User User')
    .toPromise();

  const destinationDir = getConfigDir('type');
  const destinationTestDir = getConfigDir('type_test');

  const modulePath = getModulePath(destinationDir, moduleName);
  const testPath = getTestPath(modulePath);

  const typeFilepath = path.join(modulePath, `${name}Type.js`);
  const typeTestFilepath = path.join(testPath, `${name}Type.spec.js`);

  assert.file([
    typeFilepath,
    typeTestFilepath,
  ]);

  const files = {
    type: getFileContent(path.join(folder, typeFilepath)),
    typeTest: getFileContent(path.join(folder, typeTestFilepath)),
  };

  expect(files).toMatchSnapshot();
});
