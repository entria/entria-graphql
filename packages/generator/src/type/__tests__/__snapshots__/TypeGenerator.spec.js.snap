// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate a type 1`] = `
Object {
  "type": "// @flow
import {
  GraphQLObjectType,
  GraphQLString,
} from 'graphql';
import { globalIdField } from 'graphql-relay';
import type { GraphQLObjectTypeConfig } from 'graphql';
import type { GraphQLContext } from '../../TypeDefinition';

import { NodeInterface } from '../../../graphql/interface/NodeInterface';
import Example from './ExampleLoader';

export default new GraphQLObjectType(
  ({
    name: 'Example',
    description: 'Represents Example',
    fields: () => ({
      id: globalIdField('Example'),
      example: {
        type: GraphQLString,
        description: 'My example field',
        resolve: obj => obj.example,
      },
    }),
    interfaces: () => [NodeInterface],
  }: GraphQLObjectTypeConfig<Example, GraphQLContext>),
);
",
  "typeTest": "import { graphql } from 'graphql';
import { schema } from '../../schema';
import {
  getContext,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../../test/helper';
import {
  User,
  Example,
} from '../../../models';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should retrieve a record', async () => {
  const user = await new User({
    name: 'user',
    email: 'user@example.com',
  }).save();

  // TODO: query to return a record
  const query = \`
    query Q {
      node(id:\\"123\\") {
        id
      }
    }
  \`;

  const variables = {

  };
  const rootValue = {};
  const context = getContext({ user });

  const { errors, data } = await graphql(schema, query, rootValue, context, variables);

  expect(data.node).toBe(null);
  expect(errors).toBe(undefined);
});
",
}
`;

exports[`generate a type with schema 1`] = `
Object {
  "type": "// @flow
import {
  GraphQLObjectType,
  GraphQLString,
  GraphQLList,
} from 'graphql';
import { globalIdField } from 'graphql-relay';
import type { GraphQLObjectTypeConfig } from 'graphql';
import type { GraphQLContext } from '../../TypeDefinition';

import { NodeInterface } from '../../../graphql/interface/NodeInterface';
import UserType from '../user/UserType';
import * as UserLoader from '../user/UserLoader';
import CommentType from '../comment/CommentType';
import * as CommentLoader from '../comment/CommentLoader';
import Post from './PostLoader';

export default new GraphQLObjectType(
  ({
    name: 'Post',
    description: 'Represents Post',
    fields: () => ({
      id: globalIdField('Post'),
      title: {
        type: GraphQLString,
        description: '',
        resolve: obj => obj.title,
      },
      user: {
        type: UserType,
        description: 'User that created this post',
        resolve: async (obj, args, context) => await UserLoader.load(context, obj.user),
      },
      slug: {
        type: GraphQLString,
        description: 'Used for SEO',
        resolve: obj => obj.slug,
      },
      tags: {
        type: GraphQLList(GraphQLString),
        description: '',
        resolve: obj => obj.tags,
      },
      oldSlugs: {
        type: GraphQLList(GraphQLString),
        description: 'Old slugs used by this post',
        resolve: obj => obj.oldSlugs,
      },
      comments: {
        type: GraphQLList(CommentType),
        description: '',
        resolve: async (obj, args, context) => await CommentLoader.loadCommentsByIds(context, obj.comments),
      },
      externalComments: {
        type: GraphQLList(CommentType),
        description: 'Comments from external source',
        resolve: async (obj, args, context) => await CommentLoader.loadExternalCommentsByIds(context, obj.externalComments),
      },
      createdAt: {
        type: GraphQLString,
        description: '',
        resolve: obj => obj.createdAt ? obj.createdAt.toISOString() : null,
      },
      updatedAt: {
        type: GraphQLString,
        description: '',
        resolve: obj => obj.updatedAt ? obj.updatedAt.toISOString() : null,
      },
    }),
    interfaces: () => [NodeInterface],
  }: GraphQLObjectTypeConfig<Post, GraphQLContext>),
);
",
  "typeTest": "import { graphql } from 'graphql';
import { schema } from '../../schema';
import {
  getContext,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../../test/helper';
import {
  User,
  Post,
} from '../../../models';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should retrieve a record', async () => {
  const user = await new User({
    name: 'user',
    email: 'user@example.com',
  }).save();

  // TODO: query to return a record
  const query = \`
    query Q {
      node(id:\\"123\\") {
        id
      }
    }
  \`;

  const variables = {

  };
  const rootValue = {};
  const context = getContext({ user });

  const { errors, data } = await graphql(schema, query, rootValue, context, variables);

  expect(data.node).toBe(null);
  expect(errors).toBe(undefined);
});
",
}
`;

exports[`generate a type with schema and without timestamps 1`] = `
Object {
  "type": "// @flow
import {
  GraphQLObjectType,
  GraphQLString,
  GraphQLBoolean,
} from 'graphql';
import { globalIdField } from 'graphql-relay';
import type { GraphQLObjectTypeConfig } from 'graphql';
import type { GraphQLContext } from '../../TypeDefinition';

import { NodeInterface } from '../../../graphql/interface/NodeInterface';
import User from './UserLoader';

export default new GraphQLObjectType(
  ({
    name: 'User',
    description: 'Represents User',
    fields: () => ({
      id: globalIdField('User'),
      name: {
        type: GraphQLString,
        description: '',
        resolve: obj => obj.name,
      },
      password: {
        type: GraphQLString,
        description: '',
        resolve: obj => obj.password,
      },
      email: {
        type: GraphQLString,
        description: '',
        resolve: obj => obj.email,
      },
      active: {
        type: GraphQLBoolean,
        description: '',
        resolve: obj => obj.active,
      },
      lastLoginAt: {
        type: GraphQLString,
        description: '',
        resolve: obj => obj.lastLoginAt ? obj.lastLoginAt.toISOString() : null,
      },
    }),
    interfaces: () => [NodeInterface],
  }: GraphQLObjectTypeConfig<User, GraphQLContext>),
);
",
  "typeTest": "import { graphql } from 'graphql';
import { schema } from '../../schema';
import {
  getContext,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../../test/helper';
import {
  User,
  User,
} from '../../../models';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should retrieve a record', async () => {
  const user = await new User({
    name: 'user',
    email: 'user@example.com',
  }).save();

  // TODO: query to return a record
  const query = \`
    query Q {
      node(id:\\"123\\") {
        id
      }
    }
  \`;

  const variables = {

  };
  const rootValue = {};
  const context = getContext({ user });

  const { errors, data } = await graphql(schema, query, rootValue, context, variables);

  expect(data.node).toBe(null);
  expect(errors).toBe(undefined);
});
",
}
`;
